src/Orleans/Streams/PubSub/MemoryPubSubStore.cs

```diff
@@ -0,0 +1,258 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans.Streams.PubSub
+{
+    [Serializable]
+    internal class MemoryPubSubStreamSubscriptions
+    {
+        // use ConcurrentDictionary as a concurrent HasSet
+        //http://stackoverflow.com/questions/18922985/concurrent-hashsett-in-net-framework
+        private readonly ConcurrentDictionary<PubSubPublisherState, byte> producers;
+        private readonly ConcurrentDictionary<PubSubSubscriptionState, byte> consumers;
+
+        public MemoryPubSubStreamSubscriptions()
+        {
+            producers = new ConcurrentDictionary<PubSubPublisherState, byte>();
+            consumers = new ConcurrentDictionary<PubSubSubscriptionState, byte>();
+        }
+
+        public void AddProducer(PubSubPublisherState state)
+        {
+            producers.AddOrUpdate(state, byte.MinValue, (publisherState, b) => b);
+        }
+
+        public void RemoveProducer(PubSubPublisherState state)
+        {
+            byte b;
+            bool re = producers.TryRemove(state, out b);
+            if (!re)
+            {
+                throw new OrleansException($"{this.GetType().Name} : Remove producer failed");
```

Comment:
this is copy the logic from GrainBasedPubSub on throwing exception when register/unregister failed. If we don't throw exception, I think it's better to give an return value of false , or something similar to indicate the failure. 

I think this really depend on what orleans application it is written for. If this Orleans application is mainly about streaming, then I think the user will want to get a return value from runtime to the application layer about the failure. If it is for APM, then it is less concerning to indicate success or failure here. 

I think it's the runtime's job to return the result, the method should indicate result on success or failure, more than just logging it. The application layer can always ignore the failure if too annoy.